### PR TITLE
Include last modified time in cache selector digests instead of metadata

### LINKED ISSUE #1073: Consider having cache selector include last modified time as well.
Currently, the last modified time is stored as metadata in each cache entry. We could actually have that be a part of the data that generates the selector (along with the layer entries). This will save having to retrieve and check metadata content.

### PR DESCRIPTION
**Pull request by TadCordle**
Includes the source file modification time in the generated selector digests. As a result, we no longer need the cache metadata file, which ends up simplifying a lot of stuff.

Fixes #1073

### CODE REVIEW

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 1]
```diff
@@ -63,10 +64,11 @@ public void testLayerEntryTemplate_compareTo() {
   }
 
   @Test
-  public void testToSortedJsonTemplates() {
+  public void testToSortedJsonTemplates() throws IOException {
     Assert.assertEquals(
         toLayerEntryTemplates(IN_ORDER_LAYER_ENTRIES),
-        LayerEntriesSelector.toSortedJsonTemplates(OUT_OF_ORDER_LAYER_ENTRIES));
+        ImmutableList.copyOf(
```
[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: iirc this test was failing because the `ArrayList` returned by `LayerEntriesSelector#toSortedJsonTemplates` wasn't equal to the `ImmutableList` returned by `toLayerEntryTemplates`, even though the contents were the same.

[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: Maybe we should just make `toLayerEntryTemplates` return `List`. Returning `ImmutableList` is too limiting without much value in tests.

[the remark this answers is not in the document]
Comment (coollog) [reply jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: Hmm, that's weird - `ArrayList` should be equal to `ImmutableList` if contents are the same.

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java [hunk 2]
```diff
@@ -82,24 +96,26 @@ public boolean equals(Object other) {
...
       return sourceFile.equals(otherLayerEntryTemplate.sourceFile)
-          && extractionPath.equals(otherLayerEntryTemplate.extractionPath);
+          && extractionPath.equals(otherLayerEntryTemplate.extractionPath)
+          && lastModifiedTime.equals(otherLayerEntryTemplate.lastModifiedTime);
     }
 
     @Override
     public int hashCode() {
-      return Objects.hash(sourceFile, extractionPath);
+      return Objects.hash(sourceFile, extractionPath, extractionPath);
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java]: Why is `extractionPath`  here twice?  Should the second be `lastModifiedTime`?

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 3]
```diff
@@ -63,6 +67,29 @@
         .collect(ImmutableList.toImmutableList());
   }
 
+  @Test
+  public void testLayerEntryTemplate_fileModified() throws IOException {
+    TemporaryFolder temporaryFolder = new TemporaryFolder();
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: I would recommend using `@Rule` since it would run the cleanup of the created folder automatically.

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 4]
```diff
@@ -63,6 +67,29 @@
...
+    temporaryFolder.create();
+    Path layerFile = temporaryFolder.newFolder("testFolder").toPath().resolve("file");
+    LayerEntry layerEntry = new LayerEntry(layerFile, AbsoluteUnixPath.get("/extraction/path"));
+
+    LayerEntryTemplate layerEntryTemplate = new LayerEntryTemplate(layerEntry);
+    DescriptorDigest expectedSelector =
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: This seems to be testing `LayerEntryTemplate` and duplicating the logic in `LayerEntriesSelector`. Should probably test `LayerEntriesSelector#generateSelector` instead.

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 5]
```diff
@@ -63,6 +67,29 @@
         .collect(ImmutableList.toImmutableList());
   }
 
+  @Test
+  public void testLayerEntryTemplate_fileModified() throws IOException {
+    TemporaryFolder temporaryFolder = new TemporaryFolder();
+    temporaryFolder.create();
+    Path layerFile = temporaryFolder.newFolder("testFolder").toPath().resolve("file");
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: This file should probably exist - we wouldn't ever be adding a layer entry for a file that doesn't exist.

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java (hunk 2 above)
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java]: oops

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java [hunk 6]
```diff
@@ -56,11 +60,16 @@
...
     private final String extractionPath;
+    private final Instant lastModifiedTime;
 
     @VisibleForTesting
-    LayerEntryTemplate(LayerEntry layerEntry) {
+    LayerEntryTemplate(LayerEntry layerEntry) throws IOException {
       sourceFile = layerEntry.getAbsoluteSourceFileString();
       extractionPath = layerEntry.getAbsoluteExtractionPathString();
+      lastModifiedTime =
+          Files.exists(layerEntry.getSourceFile())
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java]: Hmm, I don't think any of `LayerEntry` files should not exist. This should probably just let `Files.getLastModifiedTime` throw its exception.

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 7]
```diff
@@ -52,7 +51,15 @@
       ImmutableList<LayerEntry> layerEntries) {
     return layerEntries
         .stream()
-        .map(layerEntry -> new LayerEntryTemplate(layerEntry, Instant.EPOCH.toString()))
+        .map(
+            layerEntry -> {
+              try {
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: For this kind of method, should probably just use an iterative approach over stream to avoid the try-catch here.

**Comment by TadCordle** [comment]
I think we're covering a lot of cases already, but if you can think of more specific edge cases we should test for let me know.

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 8]
```diff
@@ -82,10 +93,25 @@ public void testGenerateSelector_empty() throws IOException {
...
+  @Test
+  public void testGenerateSelector_fileModified() throws IOException, InterruptedException {
+    Path layerFile = temporaryFolder.newFolder("testFolder").toPath().resolve("file");
+    Files.write(layerFile, "hello".getBytes(StandardCharsets.UTF_8));
+    LayerEntry layerEntry = new LayerEntry(layerFile, AbsoluteUnixPath.get("/extraction/path"));
+    DescriptorDigest expectedSelector =
+        LayerEntriesSelector.generateSelector(ImmutableList.of(layerEntry));
+
+    // Verify that changing modified time generates a different selector
+    Thread.sleep(1000);
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: `Files.setLastModified` to set the last modified time explicitly.

File: jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java [hunk 9]
```diff
@@ -82,10 +93,25 @@ public void testGenerateSelector_empty() throws IOException {
...
+    Files.write(layerFile, "goodbye".getBytes(StandardCharsets.UTF_8));
+    Assert.assertNotEquals(
+        expectedSelector, LayerEntriesSelector.generateSelector(ImmutableList.of(layerEntry)));
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/cache/LayerEntriesSelectorTest.java]: Could also just check that setting last modified back to the original generates the same selector.

File: jib-maven-plugin/CHANGELOG.md [hunk 10]
```diff
@@ -9,6 +9,7 @@ All notable changes to this project will be documented in this file.
 
 ### Changed
 
+- Last modified time of layer entries is now stored in selectors instead of metadata file in cache ([#1073](https://github.com/GoogleContainerTools/jib/issues/1073))
```
[the remark this answers is not in the document]
Comment (coollog) [reply jib-maven-plugin/CHANGELOG.md]: Yea also this feature wasn't present in the previous version so nothing really changed yet haha.

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java [hunk 11]
```diff
@@ -82,24 +93,27 @@ public boolean equals(Object other) {
...
    *
-   * @param layerEntries
+   * @param layerEntries the list of {@link LayerEntry} to convert
    * @return list of {@link LayerEntryTemplate} after sorting
+   * @throws IOException if checking the file creation time of a layer entry fails
    */
   @VisibleForTesting
-  static List<LayerEntryTemplate> toSortedJsonTemplates(List<LayerEntry> layerEntries) {
+  static List<LayerEntryTemplate> toSortedJsonTemplates(List<LayerEntry> layerEntries)
+      throws IOException {
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java]: stray exception?

Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/cache/LayerEntriesSelector.java]: Nope, `LayerEntryTemplate`'s constructor can throw `IOException` when getting last modified time.
